### PR TITLE
Feature/ASAA-23 Compose UI x Karumi Shot Testing

### PR DESCRIPTION
**Pull request by Br-DiorD**
Initial implementation of the Karumi Screenshot testing package with Compose UI testing combined.

Code uses Compose UI tests to collect the screenshots and karumi to save and compare them 

Please see the [Screenshot Testing: Compose UI x Karumi](https://confluence.bottlerocketapps.com/display/BKB/Screenshot+Testing%3A+Compose+UI+x+Karumi) Confluence page created

### CODE REVIEW

File: app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -13,7 +14,8 @@
         android:roundIcon="@mipmap/ic_launcher_round"
         android:supportsRtl="true"
         android:networkSecurityConfig="@xml/network_security_config"
-        android:theme="@style/AppTheme">
+        android:theme="@style/AppTheme"
+        android:requestLegacyExternalStorage="true">
```
Comment (BottleRocket-Colin) [inline app/src/main/AndroidManifest.xml]: I know during some of conversations we discussed that this external storage may not be needed. Have we tested that?

**Review by BottleRocket-Colin — COMMENTED** [review]
Is it possible to adjust font size so screen shots look reasonable?

**Comment by Br-DiorD** [comment]
@BottleRocket-Colin so the font size is still a pending questions for me. The tests are suppose to run as and we shouldn't have to add that but I will continue to look into solution.

File: app/src/main/AndroidManifest.xml (hunk 1 above)
Comment (Br-DiorD) [reply app/src/main/AndroidManifest.xml]: @BottleRocket-Colin I'll check now

File: app/src/androidTest/java/com/bottlerocketstudios/brarchitecture/HomeScreenUITest.kt [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+            HomeScreenState(
+                repositories = emptyList<UserRepositoryUiModel>().asMutableState(),
+                itemSelected = {}
+            )
+        )
+
+        compareScreenshot(composeRule.onRoot())
+    }
+
+    fun renderComponent(state: HomeScreenState? = null) {
```
Comment (BottleRocket-Colin) [inline app/src/androidTest/java/com/bottlerocketstudios/brarchitecture/HomeScreenUITest.kt]: @Br-DiorD Think I    figured out our issue.  We should wrap HomeScreen like this:

ArchitectureDemoTheme {
     HomeScreen(state)
}

Also, Could HomeScreenState in function signature be non null?

**Review by BottleRocket-Colin — CHANGES_REQUESTED** [review]

File: app/src/androidTest/java/com/bottlerocketstudios/brarchitecture/HomeScreenUITest.kt (hunk 2 above)
Comment (Br-DiorD) [reply app/src/androidTest/java/com/bottlerocketstudios/brarchitecture/HomeScreenUITest.kt]: @BottleRocket-Colin nice! I'll give it a shot after this meeting. I do wonder why it is only an issue on certain screenshots because the Pull RequestScreenshot shot is fine.

**Comment by jaredstockton** [comment]
@Br-DiorD I'll take a look at this tomorrow and provide some feedback!

File: buildSrc/src/main/kotlin/Dependencies.kt [hunk 3]
```diff
@@ -63,6 +67,10 @@ object Config {
         // Gradle version plugin; use dependencyUpdates task to view third party dependency updates via `./gradlew dependencyUpdates` or AS Gradle -> [project]] -> Tasks -> help -> dependencyUpdates
         // https://github.com/ben-manes/gradle-versions-plugin/releases
         const val GRADLE_VERSIONS = "com.github.ben-manes:gradle-versions-plugin:0.41.0"
+
+        // Gradle plugin and a core android library thought to run screenshot tests for Android
+        // https://github.com/pedrovgs/Shot
+        const val KARUMI_SCREENSHOT_TESTING = "com.karumi:shot:5.14.0"
```
Comment (jaredstockton) [inline buildSrc/src/main/kotlin/Dependencies.kt]: @Br-DiorD nit: Update to the latest released versions, `5.14.1` (see https://github.com/pedrovgs/Shot/releases)

File: buildSrc/src/main/kotlin/Dependencies.kt [hunk 4]
```diff
@@ -491,3 +507,9 @@ fun DependencyHandler.kotlinxCoroutineTestingDependencies() {
 fun DependencyHandler.turbineDependencies() {
     testImplementation(TestLibraries.TURBINE)
 }
+
+fun DependencyHandler.composeUITestingDependencies() {
```
Comment (jaredstockton) [inline buildSrc/src/main/kotlin/Dependencies.kt]: @Br-DiorD nit: Refactor to `composeUiTestingDependencies`

File: compose/build.gradle.kts [hunk 5]
```diff
@@ -23,7 +23,7 @@ android {
     compileOptions {
         sourceCompatibility = JavaVersion.VERSION_11
         targetCompatibility = JavaVersion.VERSION_11
-        isCoreLibraryDesugaringEnabled = true
+        isCoreLibraryDesugaringEnabled = false
```
Comment (jaredstockton) [inline compose/build.gradle.kts]: @Br-DiorD Is this necessary? Curious to hear what kind of issues you were seeing when it was enabled.

File: buildSrc/src/main/kotlin/Dependencies.kt [hunk 6]
```diff
@@ -63,6 +67,10 @@ object Config {
         // Gradle version plugin; use dependencyUpdates task to view third party dependency updates via `./gradlew dependencyUpdates` or AS Gradle -> [project]] -> Tasks -> help -> dependencyUpdates
         // https://github.com/ben-manes/gradle-versions-plugin/releases
         const val GRADLE_VERSIONS = "com.github.ben-manes:gradle-versions-plugin:0.41.0"
+
+        // Gradle plugin and a core android library thought to run screenshot tests for Android
+        // https://github.com/pedrovgs/Shot
```
Comment (jaredstockton) [inline buildSrc/src/main/kotlin/Dependencies.kt]: @Br-DiorD Thanks for adding the link!

File: app/src/androidTest/java/com/bottlerocketstudios/brarchitecture/HomeScreenUITest.kt (hunk 2 above)
Comment (jaredstockton) [inline app/src/androidTest/java/com/bottlerocketstudios/brarchitecture/HomeScreenUITest.kt]: @Br-DiorD I agree with Colin that we should change the type of the state argument to be non nullable as seen below:

```kotlin
fun renderComponent(state: HomeScreenState) {
```

File: app/src/androidTest/java/com/bottlerocketstudios/brarchitecture/PullRequestScreenshotTest.kt [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+        compareScreenshot(composeRule.onRoot())
+    }
+
+    @Test
+    fun pullRequestNoResultItemsScreenshot() {
+        renderComponent(emptyList)
+        compareScreenshot(composeRule.onRoot())
+    }
+
+    fun renderComponent(state: PullRequestScreenState? = null) {
```
Comment (jaredstockton) [inline app/src/androidTest/java/com/bottlerocketstudios/brarchitecture/PullRequestScreenshotTest.kt]: @Br-DiorD We should change the type of the state argument to be non nullable as seen below:

```kotlin
fun renderComponent(state: PullRequestScreenState) {
```

File: app/src/androidTest/java/com/bottlerocketstudios/brarchitecture/PullRequestScreenshotTest.kt [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+        composeRule.setContent {
+            if (state != null) PullRequestScreen(state)
```
Comment (jaredstockton) [inline app/src/androidTest/java/com/bottlerocketstudios/brarchitecture/PullRequestScreenshotTest.kt]: @Br-DiorD Remove this unnecessary null check (once the above comment is addressed)

File: app/src/androidTest/java/com/bottlerocketstudios/brarchitecture/HomeScreenUITest.kt [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+        composeRule.setContent {
+            if (state != null) HomeScreen(state)
```
Comment (jaredstockton) [inline app/src/androidTest/java/com/bottlerocketstudios/brarchitecture/HomeScreenUITest.kt]: @Br-DiorD Remove this unnecessary null check (once the above comment is addressed)

**Review by jaredstockton — CHANGES_REQUESTED** [review]
Note that I'm currently unable to execute the tests OR record new screenshots for my API 31 Pixel 4 emulator. I'm going to continue to look into this.

@Br-DiorD It would be awesome if there was some way to dictate the emulator configuration to be used to run the tests so any dev would be able to run the tests on their machine and get the same results. If there isn't a programmatic way to do that, even just listing the emulator configuration in the README.md would be helpful

**Comment by jaredstockton** [comment]
> Note that I'm currently unable to execute the tests OR record new screenshots for my API 31 Pixel 4 emulator. I'm going to continue to look into this.

tldr; Turns out I needed to uninstall a previously installed Architecture Demo build before attempting to run the integration tests.

**Findings:**
The Android Studio logs mentioned this error: `Unknown platform error occurred when running the UTP test suite. Please check logs for details.`
Not too helpful but there was also a reference to a pb file (`app/build/outputs/androidTest-results/connected/flavors/internal/test-result.pb`). I opened it in a text editor (Visual Studio Code, specifically) and saw that the root cause of the failure mentioned within:

```
Message: Failed to install APKs: INSTALL_FAILED_SHARED_USER_INCOMPATIBLE
```

File: app/src/androidTest/java/com/bottlerocketstudios/brarchitecture/HomeScreenUITest.kt (hunk 2 above)
Comment (BottleRocket-Colin) [reply app/src/androidTest/java/com/bottlerocketstudios/brarchitecture/HomeScreenUITest.kt]: @Br-DiorD _ would say that the PR request screenshot is probably not correct but just closer to being correct.  It's a matter of where our h4 and h5 typographies are closer to the default sizes but our h1 is very small compared to the h1 default.  So some screens will show closer to correct than others depending on which typography elements are being used.

File: compose/build.gradle.kts (hunk 5 above)
Comment (Br-DiorD) [reply compose/build.gradle.kts]: @br-jared-stockton I can't remember what the failure was but I'm trying repro it now to let you know.

File: buildSrc/src/main/kotlin/Dependencies.kt (hunk 3 above)
Comment (Br-DiorD) [reply buildSrc/src/main/kotlin/Dependencies.kt]: -@br-jared-stockton I downgraded to 5.14.0 due to an issue they were still trying to fix with 5.14.1. I'm trying to figure out what it was now.-

Comment (Br-DiorD) [reply buildSrc/src/main/kotlin/Dependencies.kt]: nvm appears fixed whatever it was yay !

File: compose/build.gradle.kts (hunk 5 above)
Comment (Br-DiorD) [reply compose/build.gradle.kts]: @br-jared-stockton can't repro the failure we made this change to fix maybe @BottleRocket-Colin can remember but I'm not sure. I'll revert for now since all seems well.

**Comment by Br-DiorD** [comment]
> 
@br-jared-stockton  Got to remind myself to add this to the confluence but yes you have to uninstall previous app version IOT get the test to run because the addition of the sharedUserId in the Manifest requires a complete reboot to implement itself i guess.

**Comment by jaredstockton** [comment]
> @Br-DiorD It would be awesome if there was some way to dictate the emulator configuration to be used to run the tests so any dev would be able to run the tests on their machine and get the same results. If there isn't a programmatic way to do that, even just listing the emulator configuration in the README.md would be helpful

@Br-DiorD This is the last thing I'd like to see done for this PR, primarily so that devs/CI can setup an emulator to get consistent results (meaning any team member could setup their emulator with some given configuration and run the snapshot tests successfully). I imagine that the emulator would match the emulator used to check in the screenshots :)

**Comment by Br-DiorD** [comment]
@br-jared-stockton on it. I don't remember having to do anything on the emulator for it to work so I'm going to create a new 1 and start from scratch to see what happens 👍

**Comment by Br-DiorD** [comment]
@br-jared-stockton updated with link to Confluence where I added Emulator setup steps. Link can be found in both the Base test file and the README.md file.

File: app/src/androidTest/java/com/bottlerocketstudios/brarchitecture/HomeScreenUITest.kt (hunk 2 above)
Comment (Br-DiorD) [reply app/src/androidTest/java/com/bottlerocketstudios/brarchitecture/HomeScreenUITest.kt]: @BottleRocket-Colin fixed. Sorry for the delay :)

**Comment by jaredstockton** [comment]
> updated with link to Confluence where I added Emulator setup steps. Link can be found in both the Base test file and the README.md file.

@Br-DiorD I was thinking of something more along the lines of the scripts that are in the shot-consumer folder of the shot dependency: https://github.com/pedrovgs/Shot/tree/master/shot-consumer/scripts

The idea is that a new developer would be able to execute `create_emulator.sh` and `start_emulator.sh` and validate all screenshots OR add new screenshot tests that match the screenshots already present (effectively the same emulator with same OS + screen width + screen height + pixel density)

You could copy/paste those initial scripts into the architecture demo project and then tweak to suite (if you want to use api 31 and different screen width/height values, etc).

Then you'd need to retake all screenshots with the emulator created from the script.

Once all of that is done, then anybody would be able to run the screenshot tests and take new ones that match the existing emulator criteria without having to manually create/configure a new emulator on their machine :D

**Review by jaredstockton — CHANGES_REQUESTED** [review]
See previous comment
